---
title: "Sweetpotato Analysis"
date: September 27, 2023
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, cache = TRUE)
```

# Introduction

The studied population comprises a collection of cultivars and three bi-parental populations: Beauregard x Regal, its reciprocal cross, and Beauregard x Uplifter. In total, 376 individuals were genotyped using the DArTag platform. The aim of this analysis is to assess the DArTag genotyping technology in a two-fold strategy: 

1) Considering the whole population, we will compute the genomic relationship matrix (G) and examine the distribution of the individuals in a Principal Component Analysis (PCA). It is expected that related individuals, i.e., full-sibs, will form clusters, while unrelated individuals will be scattered in the PCA plot. 

2) For the bi-parental populations, we will test markers for Mendelian segregation and compute the recombination fractions. 

Finally, we will build a map and reconstruct the haplotypes of the F1 offspring."


Loading packages

```{r load, results=FALSE, message=FALSE, warning=FALSE}
require(mappoly)
require(AGHmatrix)
require(factoextra)
require(plotly)
require(ggfortify)
require(tidyverse)
```

Auxiliary functions to get *Ipomea trifida* genome position and parallel computations.

```{r aux_func, results=FALSE, message=FALSE, warning=FALSE}
#### Get I. trifida genome from BT map ####
get_trifida<-function(x){
  x<-x[grepl("Tf", x)]
  w<-strsplit(x, "_")
  res<-sapply(w, function(z) {
    if(length(z) == 6)
      return(as.numeric(z[4]))
    else
      return(as.numeric(z[3]))
  }
  )
  names(res)<-x
  return(res)
}
#### Parallel functions to build maps in multiple chromosomes ####
phasing_and_hmm_rf<-function(X,
                             thres.twopt = 5,
                             thres.hmm = 3,
                             phase.number.limit = 1000){
  fl<-paste0("map_output/output_map_ch_", X$ch, ".txt")
  sink(fl)
  map<-est_rf_hmm_sequential(input.seq = X$seq,
                             start.set = 3,
                             thres.twopt = thres.twopt,
                             thres.hmm = thres.hmm,
                             twopt = X$tpt,
                             verbose = TRUE,
                             tol = 10e-2,
                             tol.final = 10e-2,
                             phase.number.limit = phase.number.limit,
                             sub.map.size.diff.limit = 10,
                             info.tail = TRUE)
  sink()
  return(map)
}
error_model<-function(X, tol = 10e-4){
  X$maps[[1]]$seq.rf <- rep(0.01, length(X$maps[[1]]$seq.rf))
  x<-est_full_hmm_with_global_error(input.map = X,
                                    error = 0.05,
                                    tol = tol,
                                    verbose = FALSE)
  return(x)
}
```

# DArTag data

```{r dart_data}
## Loading Dart allele dose file
dose.Dart <- read.csv("DSp22-7577_Allele_Dose_Report_updateID.csv",
                      skip = 7, row.names = 1)
dim(dose.Dart)
#### Six first columns of the data set
dose.Dart[,1:6] %>% glimpse(width = 80)
dat.temp <- t(dose.Dart[,-c(1:4)])
dim(dat.temp)
#### Missing data = -9
dat.temp[is.na(dat.temp)] <- -9
#### Screening out markers with more than 20% of missing data
dat.temp <- dat.temp[,-which(apply(dat.temp, 2, function(x) sum(x==-9)) > .1*nrow(dat.temp))]
#### Computing G matrix
G.mat <- AGHmatrix::Gmatrix(dat.temp, method = "VanRaden", ploidy = 6)
#### Different color for population partitions
pop.part <- character(nrow(G.mat))
pop.part[str_detect(rownames(G.mat), "BR")] <- "F1.BR"
pop.part[str_detect(rownames(G.mat), "RB")] <- "F1.RB"
pop.part[str_detect(rownames(G.mat), "BEAUREGARD")] <- "Beau"
pop.part[str_detect(rownames(G.mat), "REGAL")] <- "Regal"
pop.part[str_detect(rownames(G.mat), "CASE")] <- "F1.BU"
pop.part[pop.part==""] <- "Pop"

#### Heat map of G matrix (RB are mostly clustered)
heatmap(G.mat)
```

# Principal component analysis

```{r pca}
prin_comp2 <- prcomp(G.mat, scale = TRUE)

#### Explained variances vs dimensions
fviz_eig(prin_comp2)

#### Extracting components for PCA 3D plot
components <- prin_comp2[["x"]]
components <- data.frame(components)
components$PC2 <- -components$PC2
components$PC3 <- -components$PC3
fig <- plot_ly(components, x = ~PC1, y = ~PC2, z = ~PC3,
               color = pop.part,
               colors = RColorBrewer::brewer.pal(n = length(unique(pop.part)), name = 'Dark2'),
               sizes = 200) %>%
  add_markers(size = 12)
fig <- fig %>%
  layout(scene = list(bgcolor = "#e5ecf6"))
```

```{r, echo=FALSE, fig.height=7, fig.width=8}
fig
```

# Genetic mapping

## Parsing data

```{r parsing}
#### Genetic mapping of F1 population
f1 <- dose.Dart[,colnames(dose.Dart)[str_detect(colnames(dose.Dart), "BR") | str_detect(colnames(dose.Dart), "RB")]]
p1<- dose.Dart[,colnames(dose.Dart)[str_detect(colnames(dose.Dart), "BEAUREGARD")]]
p2<- dose.Dart[,colnames(dose.Dart)[str_detect(colnames(dose.Dart), "REGAL")]]
#### Using markers where the two replicates of Beauregard had the same dosage calling
mrk.id <- which(apply(p1, 1, function(x) length(unique(x))==1))
#### Gathering genome positiomn
genome.pos <- as.numeric(sapply(strsplit(names(mrk.id), split = "Chr|_"), function(x) x[3]))
chrom <- as.numeric(sapply(strsplit(names(mrk.id), split = "Chr|_"), function(x) x[2]))
#### Data frame form MAPpoly
DF <- cbind(snp_id = names(mrk.id),
            P1 = p1[mrk.id,1],
            P2 = p2[mrk.id],
            chrom = chrom,
            genome_pos = genome.pos,
            f1[mrk.id,])
dat.dart <- mappoly::table_to_mappoly(DF, ploidy = 6)
plot(dat.dart)
```

## Filtering

```{r, filtering}
#### Filtering ####
dat1 <- filter_missing(dat.dart, type = 'marker', filter.thres = 0.1, inter = FALSE)
dat1 <- filter_missing(dat1, type = 'individual', filter.thres = 0.1, inter = FALSE)
#### In the bi-parental populations, some individuals were filtered 
#### out due to their kinship relationship with the parents not aligning 
#### as expected, which would ideally be intermediate between the two parents
rm.ind <- c("RB3", "RB6", "RB11", "RB15", "RB22", "RB30", "RB41",
            "RB44", "RB46", "RB50", "RB52", "RB54", "RB56",
            "RB58", "BR2", "BR3", "BR5", "BR7", "BR8", "BR9",
            "BR10", "BR13", "BR20", "BR21", "BR24", "BR26",
            "BR27", "BR28", "BR30", "BR31", "BR40", "BR41")
dat1 <- filter_individuals(input.data = dat1, ind.to.remove = rm.ind, inter = FALSE)
#### Filtering out distorted markers
s.f <- filter_segregation(dat1, chisq.pval.thres = 0.05/dat1$n.mrk, inter = F)
s <- make_seq_mappoly(s.f)
plot(s)
s
```

## Pairwise recombination fraction

```{r twopts}
#### Two-points ####
tpt <- est_pairwise_rf2(s, ncpus = 7)
m <- rf_list_to_matrix(tpt, thresh.LOD.ph = 1.0, thresh.LOD.rf = 1.0)
so <- get_genomic_order(s)
plot(so)
so <- make_seq_mappoly(so)
plot(m, ord = so, fact = 5)
```

## Grouping

```{r group}
gr <- group_mappoly(m, expected.groups = 20, inter = FALSE, comp.mat = TRUE)
plot(gr)
gr
heatmap(gr$seq.vs.grouped.snp, Colv = NA)
```

## Gathering information for each linkage group

```{r LGs, results=FALSE}
#### Assembling linkage groups (order based on genome) ####
LGS <- vector("list", 15)
#### loading BT map
load("Mollinari_et_al_2020_BT_map.rda")
names(maps) <- names(LGS) <- 1:15
#### preparing linkage groups for phasing
for(i in names(LGS)){
  s.temp <- make_seq_mappoly(dat1, s$seq.mrk.names[s$chrom == i])
  tpt.temp <- est_pairwise_rf(s.temp)
  m.temp <- rf_list_to_matrix(tpt.temp)
  bt.o <- get_trifida(names(maps[[i]]))
  A <- abs(kronecker(s.temp$genome.pos, t(bt.o), FUN = "-"))
  s.new <- make_seq_mappoly(dat1, s.temp$seq.mrk.names[order(apply(A, 1, which.min))])
  LGS[[i]] <- list(seq = s.new, tpt = tpt.temp, ch = i, mat = m.temp)
}
```

```{r map_construction, eval=FALSE}
#### Parallel map construction and genotype probability computation ####
{
  cl <- parallel::makeCluster(15)
  parallel::clusterEvalQ(cl, require(mappoly))
  parallel::clusterExport(cl,  "dat1")
  MAPs.geno <- parallel::parLapply(cl, LGS, phasing_and_hmm_rf, 
                                   thres.twopt = 5, thres.hmm = 3, 
                                   phase.number.limit = 1000)
  parallel::stopCluster(cl)
}
save(MAPs.geno, dat1, file = "result_5_3_1000.rda")
{
  cl <- parallel::makeCluster(15)
  parallel::clusterEvalQ(cl, require(mappoly))
  parallel::clusterExport(cl,  "dat1")
  final.maps <- parallel::parLapply(cl, MAPs.geno, error_model, tol = 10e-3)
  genoprob <- parallel::parLapply(cl, final.maps, calc_genoprob_error, step = 1, error = 0.05)
  parallel::stopCluster(cl)
}
homologprob<-calc_homologprob(genoprob)
save(MAPs.geno, final.maps, genoprob, homologprob, dat1, file = "result_5_3_1000.rda")
#map1 <- phasing_and_hmm_rf(LGS[[2]], thres.twopt = 5, thres.hmm = 10, phase.number.limit = 1000)
map1 <- split_and_rephase(final.maps[[1]], twopt = LGS[[1]]$tpt, gap.threshold = 20, verbose = TRUE)
```

```{r plot}
load("~/repos/collaborations/sweetpotato-breeding-insight/corrected_data/result_5_3_1000.rda")
plot(m, ord = unlist(sapply(final.maps, function(x) x$info$mrk.names)), fact = 3)
plot_map_list(final.maps, col = mp_pallet3(15))
plot_genome_vs_map(final.maps, same.ch.lg = TRUE)
plot(homologprob)
for(i in 1:15){
  plot(final.maps[[i]], mrk.names = T, cex = .5, P = paste0("B_",i), Q = paste0("R_", i))  
}
```
